### ISSUE TITLE
Adding "return undefined" at the end of a function gets it unsafely DCE'd

### ISSUE THREAD

**Issue by gaearon** [rollup/rollup#4455]
### Rollup Version

1.31.1

### Operating System (or Browser)

macOS

### Link To Reproduction

The instructions are below

### Expected Behaviour

Preparation:

```
git clone https://github.com/gaearon/react.git
cd react/
git checkout rollup-bug
yarn
```

Repro:

```
yarn build react/index,react-dom/index --type=UMD_PROD --pretty
open repro.html
```

Expected behavior:

No errors in the console.

### Actual Behaviour

Actual behaivor:

```
Uncaught TypeError: Cannot read properties of null (reading 'focusedElem')
    at restoreSelection (react-dom.production.min.js:6620:56)
    at resetAfterCommit (react-dom.production.min.js:8755:5)
    at commitRootImpl (react-dom.production.min.js:22332:7)
```

in the console.

This is caused by a `prepareForCommit()` call inside `commitBeforeMutationEffects()` getting incorrectly DCE'd away even though it has a side effect. (It writes to a local variable.)

Now open `ReactDOMHostConfig.js` and uncomment line 218:

```diff
export function prepareForCommit(containerInfo: Container): Object | null {
  eventsEnabled = ReactBrowserEventEmitterIsEnabled();
  selectionInformation = getSelectionInformation();
  let activeInstance = null;
  if (enableCreateEventHandleAPI) {
    const focusedElem = selectionInformation.focusedElem;
    if (focusedElem !== null) {
      activeInstance = getClosestInstanceFromNode(focusedElem);
    }
  }
  ReactBrowserEventEmitterSetEnabled(false);
-  // return undefined;
}
```

Now

```
yarn build react/index,react-dom/index --type=UMD_PROD --pretty
open repro.html
```

does not produce console errors.

*Note I've disabled GCC in this repro so the problem is not in the GCC pass.*

**Comment by gaearon** [comment]
(Sorry that this is against a very old version. I'm filing it anyway because it seems like a rather scary bug. I can't say whether it's present in 2.x or not without upgrading which will take us a while.)

**Comment by lukastaegert** [comment]
Here is a reproduction of the issue using the 1.31.1 REPL:

https://rollupjs.org/repl/?version=1.31.1&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmZ1bmN0aW9uJTIwcHJlcGFyZUZvckNvbW1pdChjb250YWluZXJJbmZvKSUyMCU3QiU1Q24lNUN0Y29uc29sZS5sb2coJ2VmZmVjdCcpJTNCJTVDbiUyMCUyMHJldHVybiUyMHVuZGVmaW5lZCUzQiU1Q24lN0QlNUNuJTVDbnByZXBhcmVGb3JDb21taXQoKSUyMCU3QyU3QyUyMG51bGwlM0IlMjIlMkMlMjJpc0VudHJ5JTIyJTNBdHJ1ZSU3RCU1RCUyQyUyMm9wdGlvbnMlMjIlM0ElN0IlMjJmb3JtYXQlMjIlM0ElMjJlcyUyMiUyQyUyMm5hbWUlMjIlM0ElMjJteUJ1bmRsZSUyMiUyQyUyMmFtZCUyMiUzQSU3QiUyMmlkJTIyJTNBJTIyJTIyJTdEJTJDJTIyZ2xvYmFscyUyMiUzQSU3QiU3RCU3RCUyQyUyMmV4YW1wbGUlMjIlM0FudWxsJTdE

Trying with different Rollup versions, it is fixed since version 2.0.5 https://rollupjs.org/repl/?version=2.0.5&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmZ1bmN0aW9uJTIwcHJlcGFyZUZvckNvbW1pdChjb250YWluZXJJbmZvKSUyMCU3QiU1Q24lNUN0Y29uc29sZS5sb2coJ2VmZmVjdCcpJTNCJTVDbiUyMCUyMHJldHVybiUyMHVuZGVmaW5lZCUzQiU1Q24lN0QlNUNuJTVDbnByZXBhcmVGb3JDb21taXQoKSUyMCU3QyU3QyUyMG51bGwlM0IlMjIlMkMlMjJpc0VudHJ5JTIyJTNBdHJ1ZSU3RCU1RCUyQyUyMm9wdGlvbnMlMjIlM0ElN0IlMjJmb3JtYXQlMjIlM0ElMjJlcyUyMiUyQyUyMm5hbWUlMjIlM0ElMjJteUJ1bmRsZSUyMiUyQyUyMmFtZCUyMiUzQSU3QiUyMmlkJTIyJTNBJTIyJTIyJTdEJTJDJTIyZ2xvYmFscyUyMiUzQSU3QiU3RCU3RCUyQyUyMmV4YW1wbGUlMjIlM0FudWxsJTdE which itself is ancient: 2.0.5 was released in March 2020. You should REALLY upgrade, there have not only been some DCE improvements since then but a TON of bugfixes you may want to have.

**Comment by gaearon** [comment]
Ok, that’s great to know. Thank you for checking!
